### PR TITLE
Add MySQLConnectionLifetime

### PR DESCRIPTION
**Pull request by dbnski**
## A Pull Request should be associated with an Issue.

> We wish to have discussions in Issues. A single issue may be targeted by multiple PRs.
> If you're offering a new feature or fixing anything, we'd like to know beforehand in Issues,
> and potentially we'll be able to point development in a particular direction.

Related issue: https://github.com/github/orchestrator/issues/559

> Thank you! We are open to PRs, but please understand if for technical reasons we are unable to accept each and any PR

### Description

This PR adds a new configuration variable MySQLConnectionLifetime to define the maximum lifetime for persistent database connections.

### CODE REVIEW

File: go/db/db.go [hunk 1]
```diff
@@ -113,7 +117,13 @@ func openOrchestratorMySQLGeneric() (db *sql.DB, fromCache bool, err error) {
 	if config.Config.MySQLOrchestratorUseMutualTLS {
 		uri, _ = SetupMySQLOrchestratorTLS(uri)
 	}
-	return sqlutils.GetDB(uri)
+	if db, fromCache, err = sqlutils.GetDB(uri); err != nil {
+		return nil, fromCache, err
+	}
+	if config.Config.MySQLConnectionLifetime > 0 {
+		db.SetConnMaxLifetime(time.Duration(config.Config.MySQLConnectionLifetime) * time.Second)
+	}
```
Comment (shlomi-noach) [inline go/db/db.go]: This is not needed. If the `db` is returned from cache then `SetConnMaxLifetime` has already been applied.

File: go/config/config.go [hunk 2]
```diff
@@ -129,6 +129,7 @@ type Configuration struct {
 	MySQLOrchestratorReadTimeoutSeconds        int      // Number of seconds before backend mysql read operation is aborted (driver-side)
 	MySQLDiscoveryReadTimeoutSeconds           int      // Number of seconds before topology mysql read operation is aborted (driver-side). Used for discovery queries.
 	MySQLTopologyReadTimeoutSeconds            int      // Number of seconds before topology mysql read operation is aborted (driver-side). Used for all but discovery queries.
+	MySQLConnectionLifetime                    int      // Number of seconds the mysql driver will keep database connection alive before recycling it
```
Comment (shlomi-noach) [inline go/config/config.go]: Can you please rename the variable to `MySQLConnectionLifetimeSeconds`?

**Review by shlomi-noach — CHANGES_REQUESTED** [review]
Thank you! Please see comments.

File: go/config/config.go (hunk 2 above)
Comment (dbnski) [reply go/config/config.go]: Done.

File: go/db/db.go (hunk 1 above)
Comment (almeida-pythian) [inline go/db/db.go]: I'm not sure if this was intended @maciej-dobrzanski and @shlomi-noach but I noticed when trying to compile these changes I ran into an error due to a missing function. I eventually found this section on db.go file, these last few lines are missing:

```console
// ReadTimeNow reads and returns the current timestamp as string. This is an unfortunate workaround
// to support both MySQL and SQLite in all possible timezones. SQLite only speaks UTC where MySQL has
// timezone support. By reading the time as string we get the database's de-facto notion of the time,
// which we can then feed back to it.
func ReadTimeNow() (timeNow string, err error) {
	err = QueryOrchestrator(`select now() as time_now`, nil, func(m sqlutils.RowMap) error {
		timeNow = m.GetString("time_now")
		return nil
	})
	return timeNow, err
}
```

This is the error I received when building: (just the snippet below)

```console
+ go build -i -o /root/orchestrator/bin/orchestrator -ldflags '-X main.AppVersion=eb7a3b642f6e0aa83a4257ae62441571cc14a292 -X main.BuildDescribe=v3.0.12-12-geb7a3b64-dirty' ./go/cmd/orchestrator/main.go
# github.com/github/orchestrator/go/inst
go/inst/candidate_database_instance.go:42:31: undefined: db.ReadTimeNow
```


On a positive note, once I added the missing function to the file, I was able to compile these changes without any issues. I then added the following config parameters to my orchestrator config file and I am able to verify (so far), the "invalid connection" error has disappeared (we were experiencing this quite often, before we applied this latest suggested changes, thank you for finding this):

```console
"MySQLConnectTimeoutSeconds": 10, 
"MySQLOrchestratorReadTimeoutSeconds": 10, 
"MySQLDiscoveryReadTimeoutSeconds": 10, 
"MySQLTopologyReadTimeoutSeconds": 10, 
"MySQLConnectionLifetimeSeconds": 600, 
```

Comment (shlomi-noach) [reply go/db/db.go]: `db.ReadTimeNow` is a functional that was added recently. I suspect the branch was not updated properly. I just updated it.

Comment (shlomi-noach) [reply go/db/db.go]: Please remove as per comment.

Comment (almeida-pythian) [reply go/db/db.go]: Thanks @shlomi-noach :1st_place_medal:

**Comment by shlomi-noach** [comment]
Thank you all for your patience while I look into the changes.

File: go/db/db.go (hunk 1 above)
Comment (dbnski) [reply go/db/db.go]: Reverted to the original code here, as per your request

**Review by shlomi-noach — APPROVED** [review]
Thank you

**Pull request merged**